### PR TITLE
Add a core provider to expose the BatteryManager API

### PR DESCRIPTION
**Pull request by bemasc**
See https://developer.mozilla.org/en-US/docs/Web/API/Navigator/getBattery

### CODE REVIEW

File: providers/core/core.battery.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+var BatteryProvider = function(cap, dispatchEvent) {
+  "use strict";
+  this._batteryPromise = nav && nav.getBattery ? nav.getBattery() : null;
+
+  if (this._batteryPromise) {
+    this._batteryPromise.then(function(battery) {
+      this._setupListeners(battery, dispatchEvent);
+    }.bind(this));
+  }
```
Comment (soycode) [inline providers/core/core.battery.js]: Maybe worth adding an `else { console.warn("Trying to use core.battery API without client support"); }` or such.

File: providers/core/core.battery.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+};
+
+BatteryProvider.prototype.getCharging = function() {
```
Comment (soycode) [inline providers/core/core.battery.js]: `isCharging` may be a nicer name for returning a boolean.

File: providers/core/core.battery.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  "use strict";
+  var events = [
+    "chargingchange",
+    "chargingtimechange",
+    "dischargingtimechange",
+    "levelchange"
+  ];
+  events.forEach(function (eventName) {
+    battery.addEventListener(eventName, dispatchEvent.bind(this, eventName),
+        false);
```
Comment (soycode) [inline providers/core/core.battery.js]: I think `false` is the default here and could be omitted if desired.

**Comment by soycode** [comment]
Small nits but overall looks good.

File: providers/core/core.battery.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    }.bind(this));
+  }
+};
+
+BatteryProvider.prototype.getCharging = function() {
+  "use strict";
+  if (!this._batteryPromise) {
+    return PromiseCompat.resolve(true);  // Charging / plugged in
+  }
+  return this._batteryPromise.then(function(battery) {
```
Comment (willscott) [inline providers/core/core.battery.js]: if i'm interpreting this correctly, the structure here means that creating a BatteryProvider object will run the promise for the getBattery() call. Is the resulting state object dynamic, or a snapshot of battery state at the time that the getBattery() promise is made?  (i guess, is it worth triggering `nav.getBattery()` to be called again in each of these methods?)

Comment (bemasc) [inline providers/core/core.battery.js]: The resulting object is dynamic.  (I just checked at the JS console on my laptop.)

File: providers/core/core.battery.js (hunk 3 above)
Comment (bemasc) [inline providers/core/core.battery.js]: Done.

**Pull request merged**